wolfcrypt/src/wc_pkcs11.c

```diff
@@ -0,0 +1,1899 @@
+    CK_ATTRIBUTE    keyTemplate[] = {
+        { CKA_CLASS,     &pubKeyClass, sizeof(pubKeyClass) },
+        { CKA_KEY_TYPE,  &ecKeyType,   sizeof(ecKeyType)   },
+        { operation,     &ckTrue,      sizeof(ckTrue)      },
+        { CKA_EC_PARAMS, params,       0                   },
+        { CKA_EC_POINT,  NULL,         0                   }
+    };
+
+    ret = Pkcs11EccSetParams(public_key, keyTemplate, 3);
+    if (ret == 0) {
+        len = 2 + 1 + 2 * public_key->dp->size;
+        ecPoint = XMALLOC(len, NULL, DYNAMIC_TYPE_ECC);
+        if (ecPoint == NULL)
+            ret = MEMORY_E;
+    }
+    if (ret == 0) {
+        len -= 2;
+        ecPoint[0] = 0x04;
+        ecPoint[1] = len;
+        ret = wc_ecc_export_x963(public_key, ecPoint + 2, &len);
+    }
+    if (ret == 0) {
+        keyTemplate[4].pValue     = ecPoint;
+        keyTemplate[4].ulValueLen = len + 2;
+
+        rv = session->func->C_CreateObject(session->handle, keyTemplate, 5,
+                                                                     publicKey);
+        if (rv != CKR_OK)
+            ret = WC_HW_E;
+    }
+
+    if (ecPoint != NULL)
+        XFREE(ecPoint, NULL, DYNAMIC_TYPE_ECC);
+
+    return ret;
+}
+
+/**
+ * Gets the public key data from the PKCS#11 object and puts into the ECC key.
+ *
+ * @param  key      [in]  ECC public key.
+ * @param  session  [in]  Session object.
+ * @param  pubKey   [in]  ECC public key PKCS#11 object.
+ * @return  WC_HW_E when a PKCS#11 library call fails.
+ *          MEMORY_E when a memory allocation fails.
+ *          0 on success.
+ */
+static int Pkcs11GetEccPublicKey(ecc_key* key, Pkcs11Session* session,
+                                 CK_OBJECT_HANDLE pubKey)
+{
+    int ret = 0;
+    unsigned char* point = NULL;
+    int pointSz;
+    CK_ATTRIBUTE template[] = {
+      {CKA_EC_POINT, NULL_PTR, 0},
+    };
+    CK_RV rv;
+
+    rv = session->func->C_GetAttributeValue(session->handle, pubKey, template,
+                                                                             1);
+    if (rv != CKR_OK)
+        ret = WC_HW_E;
+
+    if (ret == 0) {
+        pointSz = (int)template[0].ulValueLen;
+        point = (unsigned char *)XMALLOC(pointSz, NULL, DYNAMIC_TYPE_ECC);
+        if (point == NULL)
+            ret = MEMORY_E;
+    }
+    if (ret == 0) {
+        template[0].pValue = point;
+
+        rv = session->func->C_GetAttributeValue(session->handle, pubKey,
+                                                                   template, 1);
+        if (rv != CKR_OK)
+            ret = WC_HW_E;
+    }
+
+    /* Step over the OCTET_STRING wrapper. */
+    if (ret == 0 && pointSz < key->dp->size * 2 + 3)
+        ret = ASN_PARSE_E;
+    if (ret == 0 && point[0] != 0x04)
+        ret = ASN_PARSE_E;
+    if (ret == 0 && point[1] != key->dp->size * 2 + 1)
+        ret = ASN_PARSE_E;
+
+    if (ret == 0) {
+        ret = wc_ecc_import_point_der(point + 2, pointSz - 2, ECC_CURVE_DEF,
```

Comment:
Does this mean `Pkcs11GetEccPublicKey` only supports P-256? Is there a way to get/know the curveId here and use it?

Reply:
Fixed to look up index.